### ISSUE TITLE
Need help understanding tests

### ISSUE THREAD

**Issue by billiegoose** [phestermcs/node-sjw-test#1]
I'm confused. I'm pretty sure I like what you are doing, but... these tests totally confuse me. What is `pressfs`? What is `module-stamps`?

**Comment by ghost** [comment]
Sorry about the confusion. Hope this helps.

`pressfs` takes an object hierarchy, a 'stamp', that represents directories and files, and then creates, or 'presses' (like a machine press) it onto the filesystem.

So:

```
press('base', {
  directoryA: {
    'file.txt': 'this is a text file'
  },
  directoryB: {
    'file.js': 'console.log(\'hello word\')'
  }
})
```

makes:
```
/base
  /directoryA
    file.txt
  /directoryB
    file.js
```

you can symlink real easy by prefixing key name with `$`, and then it's value is path to target directory or file:
```
press('base', {
  realdir: {
    'index.js': 'var x = 1'
  },
  $symdir: './realdir',
  $symfile: './symdir/index.js'
})
```
makes:
```
/base
  /realdir
    index.js
  /symdir -> ./realdir
  symfile -> ./symdir/index.js
```

`module-stamps` contain methods that generate modules with dependencies, and an index.js file, that when `require()d`, tracks how it's being linked and what dependencies it's linking, along with info like `__dirname`, `realpath`, `version`, etc., and that then calls all its dependents. When you run `node` on the top level index.js, it simply requires all its dependents, then calls all them, and they in turn do the same, and the output is generated. You can then see exactly whats going on with `node` with respect to where it's loading modules from, how it's internally caching them, etc.

So

```
press('base', mod('main_1', ['depA'], {
  node_modules: {
    depA: mod('depA@1.0.0')
  }
})
```

makes:
```
/base
  index.js       // this is autogen'd to require('depA') and then call an 'access()' function
  /node_modules
    /depA
      index.js   // exact same thing as above, except has no dependencies
```

and when you run `node /base/index.js` it generates something like:
```
node   : v7.2.0  NODE_PRESERVE_SYMLINKS = 0
date   : Sat, 26 Nov 2016 04:57:49 GMT
'$'    : 'C:\src\node-sjw-test\test\_layouts'
'_'    : '$\base'

require                    linked        ld/ksh  __dirname          realpath
-------------------------  ------------  ------  ----------- key -  --------
node '_\index.js'          [main_1]      LOADED  _\                 <-
 [main_1] <depA>           [depA@1.0.0]  LOADED  _\nm\depA          <-

caller                     called        ()-cnt  __dirname          realpath
-------------------------  ------------  ------  ----------- key -  --------
node '_\index.js'()        [main_1]         1    _\                 <-
  [main_1] <depA>()        [depA@1.0.0]     1    _\nm\depA          <-
```

I use `_1` instead of `@1.0.0` just to save some room and make it easier to read.

`ld/ksh` shows if the module was `LOADED` from disk, or if its instance in cache (ksh) was returned (identified by `**` instead of `LOADED`)

`()-cnt` is the number of times the module has been called, which helps see if the same module instance is being used

`key` is whats being used as the index into the module cache for that module.

`<-` under `realpath` means it's the same as `__dirname`, otherwise it will show the realpath.

**Comment by ghost** [comment]
@wmhilton Did this help?

**Comment by billiegoose** [comment]
YES! Thank you so much! I might format that response and add it to the documentation for you. :)

That's some next-level stuff! Code that generates directories full of code that is its own acceptance tests... 😮 🎉.

I realize it's been a few days, and I am *really* hoping that I haven't missed the train and the node TSC already reviewed it or something. I've put my other personal projects on hold this weekend so I can implement a package installer that demonstrates the benefits of this PR.

**Comment by zkochan** [comment]
@wmhilton you'll want to subscribe to this thread: https://github.com/nodejs/node-eps/issues/46

I am trying to take over what Paul did. Isaac has a set of requirements that should be satisfied. This project has some good details and Isaac's own solution: https://github.com/isaacs/node6-module-system-change. I think the solution might be a mixture of Paul's and Isaac's ideas.
